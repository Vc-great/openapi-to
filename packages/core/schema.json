{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "@openapi-to/define-config",
  "title": "JSON schema for @openapi-to/core define-config",
  "type": "object",
  "required": [],
  "properties": {
    "root": {
      "type": "string",
      "description": "Project root directory. Can be an absolute path, or a path relative from the location of the config file itself.",
      "default": "process.cwd()"
    },
    "input": {
      "type": "object",
      "description": "",
      "properties": {
        "path": {
          "type": "string",
          "description": "Path to be used as the input. Can be an absolute path, or a path relative from the defined root option."
        }
      }
    },
    "output": {
      "type": "object",
      "description": "",
      "properties": {
        "path": {
          "type": "string",
          "description": "Path to be used to export all generated files. Can be an absolute path, or a path relative based of the defined root option."
        },
        "clean": {
          "type": "boolean",
          "description": "Remove previous generated files and folders."
        },
        "write": {
          "type": "boolean",
          "description": "Enabled or disable the writing to the filesystem. This is being used for the playground.",
          "default": "true"
        }
      }
    },
    "hooks": {
      "type": "object",
      "description": "Hooks that can be called when a specific action is done in openapi-to. ",
      "properties": {
        "done": {
          "oneOf": [
            {
              "type": "string",
              "description": "Hook that will be called at the end of all executions. Useful for running Prettier or ESLint to use your own linting structure."
            },
            {
              "type": "array",
              "items": {
                "type": "string"
              },
              "description": "Hook that will be called at the end of all executions. Useful for running Prettier or ESLint to use your own linting structure."
            }
          ]
        }
      }
    },
    "plugins": {
      "oneOf": [
        {
          "type": "array",
          "items": {
            "type": "array",
            "prefixItems": [
              {
                "type": "string"
              }
            ],
            "items": {
              "type": "object"
            }
          }
        },
        {
          "type": "object"
        }
      ],
      "description": "Plugins"
    }
  }
}
