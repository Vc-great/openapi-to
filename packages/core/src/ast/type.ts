import type { JSDocTagStructure } from "ts-morph";
import type {
  AmbientableNodeStructure,
  ExportableNodeStructure,
  JSDocableNodeStructure,
  Structure,
} from "ts-morph";
import type {
  BindingNamedNodeStructure,
  ExclamationTokenableNodeStructure,
  KindedStructure,
  OptionalKind,
  StructureKind,
  TypedNodeStructure,
  VariableDeclarationKind,
  WriterFunction,
} from "ts-morph";

export interface VariableStatementStructureOfObject
  extends Structure,
    VariableStatementSpecificStructure,
    JSDocableNodeStructure,
    AmbientableNodeStructure,
    ExportableNodeStructure {}

interface VariableStatementSpecificStructure
  extends KindedStructure<StructureKind.VariableStatement> {
  declarationKind?: VariableDeclarationKind;
  declarations: OptionalKind<VariableDeclarationStructure>[];
}

export interface VariableDeclarationStructure
  extends Structure,
    VariableDeclarationSpecificStructure,
    BindingNamedNodeStructure,
    InitializerExpressionableNodeStructure,
    TypedNodeStructure,
    ExclamationTokenableNodeStructure {}

interface VariableDeclarationSpecificStructure
  extends KindedStructure<StructureKind.VariableDeclaration> {}

export interface InitializerExpressionableNodeStructure {
  initializer?: string | WriterFunction | ObjectStructure[];
}

export interface ObjectStructure {
  key: string;
  value: string;
  docs: JSDocStructure[];
}

export interface JSDocStructure {
  /**
   * The description of the JS doc.
   * @remarks To force this to be multi-line, add a newline to the front of the string.
   */
  description?: string;
  /** JS doc tags (ex. `&#64;param value - Some description.`). */
  tags?: OptionalKind<JSDocTagStructure>[];
}

export type JSDoc = JSDocStructure[];
