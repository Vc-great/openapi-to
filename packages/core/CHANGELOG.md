# @openapi-to/core

## 3.0.1

### Patch Changes

- fix bug

## 3.0.0

### Major Changes

- fix bug
- fix: bug
- update
- refactor

## 3.0.0-alpha.5

### Major Changes

- fix: bug

## 3.0.0-alpha.4

### Major Changes

- fix bug

## 3.0.0-alpha.3

### Major Changes

- refactor

## 3.0.0-alpha.2

### Major Changes

- fix

## 2.4.0-alpha.1

### Patch Changes

- fix bug

## 2.4.0-alpha.0

### Minor Changes

- update all

## 2.3.0

### Minor Changes

- generate swr

## 2.2.0

### Minor Changes

- fix bug

## 2.1.0

### Minor Changes

- optimize request type zod

## 2.0.0

### Major Changes

- [`e519466`](https://github.com/Vc-great/openapi-to/commit/e5194667c7416e817a498d592c357a7ae9c05f22) Thanks [@Vc-great](https://github.com/Vc-great)! - v2

- [`a39a900`](https://github.com/Vc-great/openapi-to/commit/a39a9002dda434d8a65768f55c69875ed8ad1eea) - fix bug

- 2.0

- [`367be25`](https://github.com/Vc-great/openapi-to/commit/367be252aa434487c09c4566e77792839867b509) - fix bug

- [`7fd9e58`](https://github.com/Vc-great/openapi-to/commit/7fd9e58417ef5563dedf945fbb030b70b8b09bd6) - rc

## 2.0.0-rc.5

### Major Changes

- rc

## 2.0.0-alpha.4

### Major Changes

- fix bug

## 2.0.0-alpha.3

### Major Changes

- fix bug

## 2.0.0-alpha.2

### Major Changes

- v2
