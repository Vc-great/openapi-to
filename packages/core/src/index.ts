export * from "./defineConfig";
export * from "./errors.ts";
export { Generator } from "./Generator.ts";
export { PackageManager } from "./PackageManager.ts";
// dprint-ignore
export { SchemaGenerator } from "./SchemaGenerator.ts";
export * from "./types.ts";
/*
export default build;*/
export { AST } from "./ast/ast.ts";
export { build, requestRemoteData } from "./build.ts";
export { OpenAPI } from "./openapi/OpenAPI.ts";
export { createPlugin } from "./plugins.ts";

/*



export interface _Register {}
export type Plugins = _Register;
export type OptionsPlugins = { [K in keyof Plugins]: Plugins[K]["options"] };
* */
