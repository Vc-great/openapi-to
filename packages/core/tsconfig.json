{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "types": ["vitest/globals", "bun-types"]
  },
  "include": ["src/**/*.ts", "./package.json", "vitest.config.ts", "tsup.config.ts", "../config-ts/reset.d.ts"]
}
