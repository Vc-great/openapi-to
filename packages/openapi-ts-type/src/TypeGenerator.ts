import path from "node:path";

import _ from "lodash";
import { VariableDeclarationKind } from "ts-morph";

import { modelFolderName } from "./utils/modelFolderName.ts";
import { Component } from "./Component.ts";
import { useEnumCache } from "./EnumCache.ts";
import { Schema } from "./Schema.ts";

import type { Operation } from "oas/operation";
import type OasTypes from "oas/types";
import type { OpenAPIV3 } from "openapi-types";
import type {
  ImportDeclarationStructure,
  InterfaceDeclarationStructure,
  ModuleDeclarationStructure,
  TypeAliasDeclarationStructure,
} from "ts-morph";
import type { EnumCache } from "./EnumCache.ts";
import type { Config } from "./types.ts";

type TypeStatements =
  | InterfaceDeclarationStructure
  | TypeAliasDeclarationStructure;

type ImportStatementsOmitKind = Omit<ImportDeclarationStructure, "kind">;
type InterfaceStatementsOmitKind = Omit<InterfaceDeclarationStructure, "kind">;
type ObjectSchema = OpenAPIV3.BaseSchemaObject & {
  type: "object";
};

type ResponseObject = {
  code: string;
  jsonSchema?: {
    description?: string;
    label: string;
    schema: OasTypes.SchemaObject;
    type: string | string[];
  };
};

type ComponentObject = {
  [key: string]: OpenAPIV3.ReferenceObject | OasTypes.MediaTypeObject;
};

export class TypeGenerator {
  private operation: Operation | undefined;
  private oas: Config["oas"];
  private readonly paramsZodSchema: string;
  private readonly openapi: Config["openapi"];
  private readonly ast: Config["ast"];
  private readonly pluginConfig: Config["pluginConfig"];
  private readonly openapiToSingleConfig: Config["openapiToSingleConfig"];
  private enumCache: EnumCache = useEnumCache();
  private importCache: Set<string> = new Set<string>();
  private component: Component;
  private schema: Schema;

  private readonly modelFolderName: string = modelFolderName;
  constructor(config: Config) {
    this.oas = config.oas;
    this.ast = config.ast;
    this.pluginConfig = config.pluginConfig;
    this.openapiToSingleConfig = config.openapiToSingleConfig;
    this.openapi = config.openapi;
    this.paramsZodSchema = "paramsZodSchema";
    this.component = new Component(config);
    this.schema = new Schema(config);
  }

  get currentTagName() {
    return _.camelCase(
      this.openapi &&
        this.openapi.currentTagMetadata &&
        this.openapi.currentTagMetadata.name,
    );
  }
  get nameSpaceName(): string {
    return _.upperFirst(this.currentTagName);
  }

  get upperFirstNameSpaceName(): string {
    return _.upperFirst(this.currentTagName);
  }

  build(): void {
    _.mapValues(this.openapi.pathGroupByTag, (pathGroup, tag) => {
      const typeStatements = _.chain(pathGroup)
        .map(({ path, method, tag }) => {
          this.operation = this.openapi.setCurrentOperation(path, method, tag);
          return _.chain([] as Array<TypeStatements | null>)
            .concat(this.generateParametersType())
            .concat(this.generateRequestBodyType())
            .concat(this.generateResponseType())
            .filter(Boolean)
            .value() as Array<TypeStatements>;
        })
        .flatten()
        .filter((x) => !_.isEmpty(x))
        .value();

      const filePath = path.resolve(
        this.openapiToSingleConfig.output,
        this.upperFirstNameSpaceName + ".ts",
      );

      const refKey = [...this.openapi.refCache.keys()];

      this.ast.createSourceFile(filePath, {
        statements: [
          ...this.generateImport(refKey),
          this.generateNameSpace(typeStatements),
        ],
      });

      this.openapi.resetRefCache();
    });
    this.component.generateComponentType();
    this.generateEnum();
    this.component.generateModelIndex();
  }

  /**
   * @example
   * ```
   * export const enum TaskStatus {
   *     WAIT_START = 'WAIT_START',
   *     PROCESSING = 'PROCESSING',
   *     FINISHED = 'FINISHED',
   * }
   *
   * export const enum TaskStatusLabel {
   *     WAIT_START:'WAIT_START',
   *     PROCESSING = 'PROCESSING',
   *     FINISHED = 'FINISHED',
   * }
   *
   * export const taskStatusOption = [
   *     { label: TaskStatusLabel.WAIT_START, value: TaskStatus.WAIT_START },
   *     { label: TaskStatusLabel.PROCESSING, value: TaskStatus.PROCESSING },
   *     { label: TaskStatusLabel.FINISHED, value: TaskStatus.FINISHED },
   * ];
   * ```
   */
  generateEnum(): void {
    //
    const labelStatements = _.chain(this.enumCache.entries())
      .map(([schema, name]) => {
        return this.ast.generateVariableStatements({
          declarationKind: VariableDeclarationKind.Const,
          isExported: true,
          declarations: [
            {
              name: _.upperFirst(_.camelCase(name)) + "Label",
              initializer: this.ast.generateObject(
                ((schema.enum || []) as Array<string>).reduce(
                  (obj, item: string) => {
                    const key: string = _.upperFirst(_.camelCase(item));
                    obj[key] = "''";
                    return obj;
                  },
                  {} as { [key: string]: string },
                ),
              ),
            },
          ],
          docs: [{ description: schema.description || "" }],
        });
      })
      .value();

    //enum
    const valueStatements = _.chain(this.enumCache.entries())
      .map(([schema, name]) => {
        return this.ast.generateEnumStatement({
          isConst: true,
          name: _.upperFirst(_.camelCase(name)),
          members: (schema.enum || []).map((item: string) => {
            return {
              name: _.upperFirst(_.camelCase(item)),
              value: item,
            };
          }),
          docs: [
            {
              description: schema.description,
            },
          ],
        });
      })
      .value();

    // option
    const optionStatements = _.chain(this.enumCache.entries())
      .map(([schema, name]) => {
        return this.ast.generateVariableStatements({
          declarationKind: VariableDeclarationKind.Const,
          isExported: true,
          declarations: [
            {
              name: name + "Option",
              initializer:
                "[" +
                (schema.enum || []).reduce((arr, item: string) => {
                  const obj = this.ast.generateObject({
                    label:
                      _.upperFirst(_.camelCase(name)) +
                      "Label" +
                      "." +
                      _.upperFirst(_.camelCase(item)),
                    value:
                      _.upperFirst(_.camelCase(name)) +
                      "." +
                      _.upperFirst(_.camelCase(item)),
                  });
                  return arr + (arr ? "," : "") + obj;
                }, "") +
                "]",
            },
          ],
          docs: [{ description: schema.description || "" }],
        });
      })
      .value();

    const enumPath = path.resolve(
      this.openapiToSingleConfig.output || "",
      this.modelFolderName,
      "enum" + ".ts",
    );

    return this.ast.createSourceFile(enumPath, {
      statements: [...labelStatements, ...valueStatements, ...optionStatements],
    });
  }

  /**
   *
   * @example
   * ```
   * // ts type
   * import type { DrmsDynamicDataType } from './drmsDynamicDataZod'
   * ```
   */
  generateImport(
    importModel: Array<string>,
  ): Array<ImportDeclarationStructure> {
    const model = _.chain(importModel)
      .map(($ref: string) => _.upperFirst(this.openapi.getRefAlias($ref)))
      .value();

    const typeModel: ImportStatementsOmitKind = {
      namedImports: [...model],
      isTypeOnly: true,
      moduleSpecifier: "./" + this.modelFolderName,
    };

    const statements = _.chain([] as Array<ImportStatementsOmitKind>)
      .push(typeModel)
      .filter(Boolean)
      .value();

    return this.ast.generateImportStatements(statements);
  }

  /**
   * @example
   * ```
   * export namespace Tag {
   *  export type queryRequest = object
   * }
   * ```
   */
  generateNameSpace(
    typeStatements: Array<TypeStatements>,
  ): ModuleDeclarationStructure {
    return this.ast.generateModuleStatements({
      docs: [{ description: this.openapi.currentTagMetadata?.description }],
      isExported: true,
      name: this.nameSpaceName,
      statements: typeStatements,
    });
  }

  generateParametersType(): Array<
    InterfaceDeclarationStructure | TypeAliasDeclarationStructure
  > {
    const queryStatements = this.ast.generateInterfaceStatements({
      isExported: true,
      name: this.openapi.upperFirstRequestName + "QueryParams",
      docs: [{ description: "queryParams" }],
      properties: this.schema.getBaseTypeFromSchema(
        this.openapi.parameter?.getParametersSchema("query") || null,
      ),
    });

    const pathStatements = this.ast.generateInterfaceStatements({
      isExported: true,
      name: this.openapi.upperFirstRequestName + "PathParams",
      docs: [{ description: "pathParams" }],
      properties: this.schema.getBaseTypeFromSchema(
        this.openapi.parameter?.getParametersSchema("path") || null,
      ),
    });
    //[mediaType, bodySchema, description]

    return _.chain(
      [] as Array<
        InterfaceDeclarationStructure | TypeAliasDeclarationStructure
      >,
    )
      .concat(this.openapi.parameter?.hasQueryParameters ? queryStatements : [])
      .concat(this.openapi.parameter?.hasPathParameters ? pathStatements : [])
      .filter(Boolean)
      .value();
  }

  generateResponseType(): Array<
    InterfaceDeclarationStructure | TypeAliasDeclarationStructure
  > {
    const codes = this.openapi.response?.getResponseStatusCodes || [];

    const successCode = (codes || []).filter((code) =>
      /^(2[0-9][0-9]|300)$/.test(code),
    );
    const errorCode = (codes || []).filter((code) =>
      /^([3-5][0-9][0-9])$/.test(code),
    );
    const isSuccessCode = (code: string) => /^(2[0-9][0-9]|300)$/.test(code);

    const responseObject = _.chain([...successCode, ...errorCode])
      .map((code) => {
        return {
          code,
          jsonSchema:
            _.head(this.openapi.response?.getResponseAsJSONSchema(code)) ||
            null,
        };
      })
      .filter((x) => {
        if (isSuccessCode(x.code)) {
          return !_.isNull(x.jsonSchema);
        }
        return true;
      })
      .value() as Array<ResponseObject>;

    const errorType = this.ast.generateTypeAliasStatements({
      name: this.openapi.upperFirstRequestName + "ErrorResponse",
      type:
        _.chain(errorCode)
          .map((code) => this.openapi.upperFirstRequestName + "Response" + code)
          .join("|")
          .value() || "unknown",
      docs: [{ description: "" }],
      isExported: true,
    });

    const successDefaultType = this.ast.generateTypeAliasStatements({
      name: this.openapi.upperFirstRequestName + "Response",
      type: "unknown",
      docs: [{ description: "" }],
      isExported: true,
    });

    return _.chain(responseObject)
      .map((responseObject) =>
        this.generateResponseSingleSchema(responseObject),
      )
      .concat(errorType)
      .concat(_.isEmpty(successCode) ? successDefaultType : [])
      .filter(Boolean)
      .value();
  }

  generateResponseSingleSchema({
    code,
    jsonSchema,
  }: ResponseObject):
    | InterfaceDeclarationStructure
    | TypeAliasDeclarationStructure {
    const schema = jsonSchema?.schema;
    const description = jsonSchema?.description;
    const isError = /^([3-5][0-9][0-9])$/.test(code);
    const name =
      this.openapi.upperFirstRequestName + "Response" + (isError ? code : "");

    if (!schema) {
      return this.ast.generateTypeAliasStatements({
        name,
        type: "unknown",
        docs: [{ description: "" }],
        isExported: true,
      });
    }

    if (this.openapi.isReference(schema)) {
      return this.ast.generateTypeAliasStatements({
        name,
        type: _.upperFirst(this.openapi.getRefAlias(schema.$ref)),
        docs: [{ description: description || "" }],
        isExported: true,
      });
    }

    if (schema.type === "object") {
      return this.ast.generateInterfaceStatements({
        isExported: true,
        name,
        docs: [{ description: description || "" }],
        properties: this.schema.getBaseTypeFromSchema(schema),
      });
    }

    return this.ast.generateTypeAliasStatements({
      name,
      type: this.schema.formatterSchemaType(schema),
      docs: [{ description: description || "" }],
      isExported: true,
    });
  }

  //SchemaObject
  generateRequestBodyType(): Array<TypeStatements> | null {
    const _bodySchema = this.openapi.requestBody?.getRequestBodySchema();
    if (_bodySchema === undefined || _.isBoolean(_bodySchema)) {
      return null;
    }
    let schema: OasTypes.SchemaObject | null = null;
    if ("schema" in _bodySchema) {
      schema = _bodySchema.schema || (null as OasTypes.SchemaObject | null);
    }

    if (_.isArray(_bodySchema)) {
      schema = _.get(
        _bodySchema,
        "[1].schema",
        null,
      ) as OasTypes.SchemaObject | null;
    }

    if (schema === null) {
      return null;
    }

    if (this.openapi.isReference(schema)) {
      return [
        this.ast.generateTypeAliasStatements({
          name: this.openapi.upperFirstRequestName + "BodyParams",
          type: _.upperFirst(this.openapi.getRefAlias(schema.$ref)),
          docs: [{ description: "" }],
          isExported: true,
        }),
      ];
    }

    if (schema.type === "array") {
      return [
        this.ast.generateTypeAliasStatements({
          name: this.openapi.upperFirstRequestName + "BodyParams",
          type: this.schema.formatterSchemaType(schema),
          docs: [{ description: "" }],
          isExported: true,
        }),
      ];
    }

    return [
      this.ast.generateInterfaceStatements({
        isExported: true,
        name: this.openapi.upperFirstRequestName + "BodyParams",
        docs: [{ description: "bodyParams" }],
        properties: this.schema.getBaseTypeFromSchema(schema),
      }),
    ];
  }
}
