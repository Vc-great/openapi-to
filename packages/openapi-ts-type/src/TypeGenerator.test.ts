import path from "node:path";
import { dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { AST, OpenAPI } from "@openapi-to/core";

import _ from "lodash";
import Oas from "oas";

import petStore from "../mock/petstore.json";
import { TypeGenerator } from "./TypeGenerator.ts";

import type { OpenapiToSingleConfig } from "@openapi-to/core";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe("typerGenerator", async () => {
  const openapiToSingleConfig: OpenapiToSingleConfig = {
    input: {
      path: "",
      name: "",
    },
    output: path.resolve(__dirname, "../gen"),
    plugins: [],
  };

  const pluginConfig = {
    createZodDecorator: true,
  };
  const context = {
    output: "",
  };

  test("TypeGenerator getFullText", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    const oas = new Oas(petStore);
    const openapi = new OpenAPI({}, oas);

    const typerGenerator = new TypeGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    typerGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });
});
