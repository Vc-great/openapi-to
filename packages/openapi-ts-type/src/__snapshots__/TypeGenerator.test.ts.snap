// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`typerGenerator > TypeGenerator getFullText 1`] = `
"import type { TestDto2, Test32145, ApiResponse, Pet } from "./models";

/** Everything about your Pets */
export namespace Pet {
    /** OK */
    export type TestPostResponse = TestDto2;
    /** */
    export type TestPostResponse401 = unknown;
    /** */
    export type TestPostResponse403 = unknown;
    /** */
    export type TestPostResponse404 = unknown;
    /** */
    export type TestPostErrorResponse = TestPostResponse401 | TestPostResponse403 | TestPostResponse404;
    /** OK */
    export type TestPutResponse = TestDto2;
    /** */
    export type TestPutResponse401 = unknown;
    /** */
    export type TestPutResponse403 = unknown;
    /** */
    export type TestPutResponse404 = unknown;
    /** */
    export type TestPutErrorResponse = TestPutResponse401 | TestPutResponse403 | TestPutResponse404;
    /** */
    export type DelByTestBodyParams = Array<number>;
    /** OK */
    export type DelByTestResponse = Test32145;
    /** */
    export type DelByTestResponse401 = unknown;
    /** */
    export type DelByTestResponse403 = unknown;
    /** */
    export type DelByTestErrorResponse = DelByTestResponse401 | DelByTestResponse403;

    /** queryParams */
    export interface TestIdGetQueryParams {
        /** */
        fields?: Array<string>;
        /** */
        page: number;
        /** */
        size: number;
    }

    /** pathParams */
    export interface TestIdGetPathParams {
        /** */
        testId: number;
        /** */
        testId2: string;
    }

    /** OK */
    export type TestIdGetResponse = TestDto2;
    /** */
    export type TestIdGetResponse401 = unknown;
    /** */
    export type TestIdGetResponse403 = unknown;
    /** */
    export type TestIdGetResponse404 = unknown;
    /** */
    export type TestIdGetErrorResponse = TestIdGetResponse401 | TestIdGetResponse403 | TestIdGetResponse404;

    /** pathParams */
    export interface UploadImagePostPathParams {
        /** */
        petId: number;
    }

    /** bodyParams */
    export interface UploadImagePostBodyParams {
        /** Additional data to pass to server */
        additionalMetadata?: string;
        /** file to upload */
        file?: string;
    }

    /** successful operation */
    export type UploadImagePostResponse = ApiResponse;
    /** */
    export type UploadImagePostErrorResponse = unknown;
    /** */
    export type CreateResponse405 = unknown;
    /** */
    export type CreateErrorResponse = CreateResponse405;
    /** */
    export type CreateResponse = unknown;
    /** */
    export type UpdateResponse400 = unknown;
    /** */
    export type UpdateResponse404 = unknown;
    /** */
    export type UpdateResponse405 = unknown;
    /** */
    export type UpdateErrorResponse = UpdateResponse400 | UpdateResponse404 | UpdateResponse405;
    /** */
    export type UpdateResponse = unknown;

    /** queryParams */
    export interface FindByStatusGetQueryParams {
        /** */
        status: Array<string>;
    }

    /** successful operation */
    export type FindByStatusGetResponse = Pet;
    /** */
    export type FindByStatusGetResponse400 = unknown;
    /** */
    export type FindByStatusGetErrorResponse = FindByStatusGetResponse400;

    /** queryParams */
    export interface FindByTagsGetQueryParams {
        /** */
        tags: Array<string>;
    }

    /** successful operation */
    export type FindByTagsGetResponse = Pet;
    /** */
    export type FindByTagsGetResponse400 = unknown;
    /** */
    export type FindByTagsGetErrorResponse = FindByTagsGetResponse400;

    /** pathParams */
    export interface FindByPetIdPathParams {
        /** */
        petId: number;
    }

    /** successful operation */
    export type FindByPetIdResponse = Pet;
    /** */
    export type FindByPetIdResponse400 = unknown;
    /** */
    export type FindByPetIdResponse404 = unknown;
    /** */
    export type FindByPetIdErrorResponse = FindByPetIdResponse400 | FindByPetIdResponse404;

    /** pathParams */
    export interface PetIdPostPathParams {
        /** */
        petId: number;
    }

    /** bodyParams */
    export interface PetIdPostBodyParams {
        /** Updated name of the pet */
        name?: string;
        /** Updated status of the pet */
        status?: string;
    }

    /** */
    export type PetIdPostResponse405 = unknown;
    /** */
    export type PetIdPostErrorResponse = PetIdPostResponse405;
    /** */
    export type PetIdPostResponse = unknown;

    /** pathParams */
    export interface DelByPetIdPathParams {
        /** */
        petId: number;
    }

    /** */
    export type DelByPetIdResponse400 = unknown;
    /** */
    export type DelByPetIdResponse404 = unknown;
    /** */
    export type DelByPetIdErrorResponse = DelByPetIdResponse400 | DelByPetIdResponse404;
    /** */
    export type DelByPetIdResponse = unknown;
}

import type { Order } from "./models";

/** Access to Petstore orders */
export namespace Store {
    /** */
    export type CreateBodyParams = Order;
    /** successful operation */
    export type CreateResponse = Order;
    /** */
    export type CreateResponse400 = unknown;
    /** */
    export type CreateErrorResponse = CreateResponse400;

    /** pathParams */
    export interface FindByOrderIdPathParams {
        /** */
        orderId: number;
    }

    /** successful operation */
    export type FindByOrderIdResponse = Order;
    /** */
    export type FindByOrderIdResponse400 = unknown;
    /** */
    export type FindByOrderIdResponse404 = unknown;
    /** */
    export type FindByOrderIdErrorResponse = FindByOrderIdResponse400 | FindByOrderIdResponse404;

    /** pathParams */
    export interface DelByOrderIdPathParams {
        /** */
        orderId: number;
    }

    /** */
    export type DelByOrderIdResponse400 = unknown;
    /** */
    export type DelByOrderIdResponse404 = unknown;
    /** */
    export type DelByOrderIdErrorResponse = DelByOrderIdResponse400 | DelByOrderIdResponse404;
    /** */
    export type DelByOrderIdResponse = unknown;

    /** successful operation */
    export interface InventoryGetResponse {
    }

    /** */
    export type InventoryGetErrorResponse = unknown;
}

import type { User } from "./models";

/** Operations about user */
export namespace User {
    /** */
    export type CreateWithArrayPostErrorResponse = unknown;
    /** */
    export type CreateWithArrayPostResponse = unknown;
    /** */
    export type CreateWithListPostErrorResponse = unknown;
    /** */
    export type CreateWithListPostResponse = unknown;

    /** pathParams */
    export interface FindByUsernamePathParams {
        /** */
        username: string;
    }

    /** successful operation */
    export type FindByUsernameResponse = User;
    /** */
    export type FindByUsernameResponse400 = unknown;
    /** */
    export type FindByUsernameResponse404 = unknown;
    /** */
    export type FindByUsernameErrorResponse = FindByUsernameResponse400 | FindByUsernameResponse404;

    /** pathParams */
    export interface UsernamePutPathParams {
        /** */
        username: string;
    }

    /** */
    export type UsernamePutBodyParams = User;
    /** */
    export type UsernamePutResponse400 = unknown;
    /** */
    export type UsernamePutResponse404 = unknown;
    /** */
    export type UsernamePutErrorResponse = UsernamePutResponse400 | UsernamePutResponse404;
    /** */
    export type UsernamePutResponse = unknown;

    /** pathParams */
    export interface DelByUsernamePathParams {
        /** */
        username: string;
    }

    /** */
    export type DelByUsernameResponse400 = unknown;
    /** */
    export type DelByUsernameResponse404 = unknown;
    /** */
    export type DelByUsernameErrorResponse = DelByUsernameResponse400 | DelByUsernameResponse404;
    /** */
    export type DelByUsernameResponse = unknown;

    /** queryParams */
    export interface LoginGetQueryParams {
        /** */
        username: string;
        /** */
        password: string;
    }

    /** successful operation */
    export interface LoginGetResponse {
    }

    /** */
    export type LoginGetResponse400 = unknown;
    /** */
    export type LoginGetErrorResponse = LoginGetResponse400;
    /** */
    export type LogoutGetErrorResponse = unknown;
    /** */
    export type LogoutGetResponse = unknown;
    /** */
    export type CreateBodyParams = User;
    /** */
    export type CreateErrorResponse = unknown;
    /** */
    export type CreateResponse = unknown;
}

import type { Testdata } from "./Testdata";

/** */
export interface Test32145 {
    /** description */
    code?: number;
    /** */
    meta?: Testdata;
}

/** */
export interface Test3214 {
    /** 默认值 */
    columnDefault?: string;
    /** description */
    columnLength?: number;
    /** description */
    columnName?: string;
    /** description */
    columnRemark?: string;
    /** description */
    columnScale?: number;
    /** description */
    columnType?: string;
    /** description */
    delFlag?: boolean;
    /** */
    formId?: number;
    /** id */
    id?: number;
    /** description */
    notNull?: boolean;
}

/** */
export interface Testdata {
    /** description */
    number?: number;
    /** description */
    numberOfElements?: number;
    /** description */
    totalElements?: number;
    /** description */
    totalPages?: number;
}

/** */
export interface Sort {
    /** */
    empty?: boolean;
    /** */
    sorted?: boolean;
    /** */
    unsorted?: boolean;
}

import type { Test3214 } from "./Test3214";

/** */
export interface TestDto3 {
    /** */
    test3214?: Test3214;
}

import type { TestDto3 } from "./TestDto3";
import type { Testdata } from "./Testdata";

/** */
export interface TestDto2 {
    /** */
    content?: TestDto3;
    /** */
    meta?: Testdata;
}

/** */
export interface Test321 {
    /** */
    id?: number;
}

import type { Test321 } from "./Test321";

/** */
export interface TestDto {
    /** test321 */
    test321?: Test321;
}

import type { ApiResponse } from "./ApiResponse";

/** */
export interface RefList {
    /** */
    content?: ApiResponse;
    /** string */
    string?: string;
    /** */
    name?: string;
    /** title */
    title?: string;
    /** type */
    type?: string;
}

/** */
export interface ApiResponse {
    /** */
    code?: number;
    /** */
    type?: string;
    /** */
    message?: string;
}

/** */
export interface Category {
    /** */
    id?: number;
    /** */
    name?: string;
}

import type { Category } from "./Category";
import type { Tag } from "./Tag";

/** */
export interface Pet {
    /** */
    id?: number;
    /** */
    category?: Category;
    /** */
    name: string;
    /** */
    photoUrls: Array<string>;
    /** */
    tags?: Tag;
    /** pet status in the store */
    status?: string;
}

/** */
export interface Tag {
    /** */
    id?: number;
    /** */
    name?: string;
}

/** */
export interface Order {
    /** */
    id?: number;
    /** */
    petId?: number;
    /** */
    quantity?: number;
    /** */
    shipDate?: string;
    /** Order Status */
    status?: string;
    /** */
    complete?: boolean;
}

/** */
export interface User {
    /** */
    id?: number;
    /** */
    username?: string;
    /** */
    firstName?: string;
    /** */
    lastName?: string;
    /** */
    email?: string;
    /** */
    password?: string;
    /** */
    phone?: string;
    /** User Status */
    userStatus?: number;
}

/** description */
export const ColumnTypeLabel = {
    Char: '',
    Char2: '',
    Char3: ''
};
/** pet status in the store */
export const StatusLabel = {
    Available: '',
    Pending: '',
    Sold: ''
};
/** Order Status */
export const Status1Label = {
    Placed: '',
    Approved: '',
    Delivered: ''
};

/** description */
const enum ColumnType {
    Char = "CHAR",
    Char2 = "CHAR2",
    Char3 = "CHAR3"
}

/** pet status in the store */
const enum Status {
    Available = "available",
    Pending = "pending",
    Sold = "sold"
}

/** Order Status */
const enum Status1 {
    Placed = "placed",
    Approved = "approved",
    Delivered = "delivered"
}

/** description */
export const columnTypeOption = [{
    label: ColumnTypeLabel.Char,
    value: ColumnType.Char
}, {
    label: ColumnTypeLabel.Char2,
    value: ColumnType.Char2
}, {
    label: ColumnTypeLabel.Char3,
    value: ColumnType.Char3
}];
/** pet status in the store */
export const statusOption = [{
    label: StatusLabel.Available,
    value: Status.Available
}, {
    label: StatusLabel.Pending,
    value: Status.Pending
}, {
    label: StatusLabel.Sold,
    value: Status.Sold
}];
/** Order Status */
export const status1Option = [{
    label: Status1Label.Placed,
    value: Status1.Placed
}, {
    label: Status1Label.Approved,
    value: Status1.Approved
}, {
    label: Status1Label.Delivered,
    value: Status1.Delivered
}];

export * from "./enum";
export * from "./Test32145";
export * from "./Test3214";
export * from "./Testdata";
export * from "./Sort";
export * from "./TestDto3";
export * from "./TestDto2";
export * from "./Test321";
export * from "./TestDto";
export * from "./RefList";
export * from "./ApiResponse";
export * from "./Category";
export * from "./Pet";
export * from "./Tag";
export * from "./Order";
export * from "./User";
"
`;
